### PR TITLE
Add granular settings defaults (`defaults.settings.{local, defined, all}`)

### LINKED ISSUE #261: Fine-grained `settings` defaults
In an internal project, we allow specifying defaults for all local packages:

image

In addition, haskell-flake's default has a default that specifies a default for both global packages (`globalSettings`) and all-local packages (`localSettings`), where all-local = local packages of all haskell-flake projects:

https://github.com/srid/haskell-flake/blob/d128c7329bfc73c3eeef90f6d215d0ccd7baf78c/nix/modules/project/defaults.nix#L76-L104

Can we generalize `default-settings` from our internal project to provide defaults for these 3 cases? Perhaps:

```nix
{
  defaults.settings = {
    default-local = ... # Defaults for packages local to current project.
    default-all-local = ... # Defaults for all packages managed by haskell-flake
    default-all = ... # Defaults for all packages specified by `packages` option
  };
}
```

We can make the naming less confusing as well.

### PR DESCRIPTION
**Pull request by shivaraj-bh**
resolves #261 

`defaults.settings.default-{current, defined, all}` replaces `defaults.settings.default`

TODO:
- [x] Add tests

### CODE REVIEW

**Review by srid — COMMENTED** [review]
Nice. Quick review.

- [x] A separate `./test/*/flake.nix` would be good for testing defaults.
- [x] Rebase new `master` (#276), and add documentation in `docs/defaults.md`
  - Let's elaborate the examples in the Markdown docs, and have the option `description` link to them.
- [ ] Not sure about the `defaults.settings.default-foo` naming; let me think about it.

**Comment by srid** [comment]
cf. https://github.com/srid/haskell-flake/discussions/301

**Comment by shivaraj-bh** [comment]
added tests: https://github.com/srid/haskell-flake/pull/275/commits/18581246feea93c7d37c02787ad879d1c110b1db

**Review by srid — COMMENTED** [review]
My CI is down, but running `nixci build` locally fails:

```
$ nixci build https://github.com/srid/haskell-flake/pull/275
[..]
$ nix log /nix/store/n1qm2m8qmg3d7q25swl3h8cqxhsng9na-haskell-flake-test-0.1.0.0.drv | tail -6
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/n1qm2m8qmg3d7q25swl3h8cqxhsng9na-haskell-flake-test-0.1.0.0.drv^*'
Configuring haskell-flake-test-0.1.0.0...
CallStack (from HasCallStack):
  withMetadata, called at libraries/Cabal/Cabal/src/Distribution/Simple/Utils.hs:368:14 in Cabal-3.10.1.0:Distribution.Simple.Utils
Error: Setup: Encountered missing or private dependencies:
hello
```

**Comment by shivaraj-bh** [comment]
> My CI is down, but running nixci build locally fails:

I fixed and tested on Linux.

`hello` wasn’t a great example to use anyway, it isn’t a library package. I switched over to `random` package.

File: nix/modules/project/settings/default.nix [hunk 1]
```diff
@@ -88,6 +96,6 @@ in
               fns
           );
       in
-      lib.mapAttrs applySettingsFor project.config.settings;
+      lib.mapAttrs applySettingsFor (lib.recursiveUpdate emptyPackageSettings project.config.settings);
```
Comment (srid) [inline nix/modules/project/settings/default.nix:99]: Why do we need `lib.recursiveUpdate emptyPackageSetting`?

**Review by srid — COMMENTED** [review]
LGTM, just not sure what `emptyPackageSettings` is for exactly. When I replace it with `{}`, the test still passes.

File: nix/modules/project/settings/default.nix (hunk 1 above)
Comment (shivaraj-bh) [reply nix/modules/project/settings/default.nix:99]: It considers the packages that are mentioned in `haskellProjects.<prj-name>.packages` but not in `haskellProjects.<prj-name>.settings`. i should include a test that would cover this case. Also, is the name misleading? Anything better you suggest?

Comment (srid) [reply nix/modules/project/settings/default.nix:99]: The `settings` option already does this:

https://github.com/srid/haskell-flake/blob/847292fc793a5c15c873e52e7751ee4267ef32a0/nix/modules/project/settings/default.nix#L21-L27

So, `project.config.settings` is already polyfilled with packages from `project.config.packages`.

**Review by srid — APPROVED** [review]

**Pull request merged**